stored_requests/fetcher.go

```diff
@@ -19,52 +19,115 @@ type Fetcher interface {
 
 // Cache is an intermediate layer which can be used to create more complex Fetchers by composition.
 // Implementations must be safe for concurrent access by multiple goroutines.
-// To add a Cache layer in front of the Fetcher, see WithCache()
+// To add a Cache layer in front of a Fetcher, see WithCache()
 type Cache interface {
-	// GetRequests works much like Fetcher.FetchRequests, with a few exceptions:
+	// Get works much like Fetcher.FetchRequests, with a few exceptions:
 	//
 	// 1. Any errors should be logged by the implementation, rather than returned.
 	// 2. The returned map _may_ be written to.
 	// 3. The returned map must _not_ contain keys unless they were present in the argument ID list.
 	// 4. Callers _should not_ assume that the returned map contains a key for every id.
 	//    The returned map will miss entries for keys which don't exist in the cache.
-	GetRequests(ctx context.Context, ids []string) map[string]json.RawMessage
+	Get(ctx context.Context, ids []string) map[string]json.RawMessage
 
-	// SaveRequests stores some data in the cache. The map is from ID to the cached value.
-	//
-	// This is a best-effort method. If the cache call fails, implementations should log the error.
-	SaveRequests(ctx context.Context, values map[string]json.RawMessage)
+	// Invalidate will ensure that all values associated with the given IDs
+	// are no longer returned by the cache until new values are saved via Update
+	Invalidate(ctx context.Context, ids []string)
+
+	// Update will update the given values in the cache
+	Update(ctx context.Context, data map[string]json.RawMessage)
+}
+
+// CacheableFetcher is a type of Fetcher that can be composed with one or more
+// Caches via WithCache and provide updates and validations to those Caches
+// when necessary
+type CacheableFetcher interface {
```

Comment:
I was thinking more on this and I think the invalidation shouldn't be the responsibility of the Fetcher to bubble up as it is in the PR currently. Instead, we should probably have a separate simple interface for an event producer (let's call it `EventProducer` for now for lack of creativity) so that `Fetcher`s can just worry about fetching. Then composition could be done like:

![stored_request_cache](https://user-images.githubusercontent.com/9155109/37424545-145404c2-2797-11e8-8583-11a70878af18.png)

Reply:
This looks like a good architecture to me. Also more flexible, so it could easily support a cache invalidation driven from something other the data source, like an explicit cache invalidation API. That would be useful for datasources that don't support events.